### PR TITLE
feat: Expand environment variables in array setting values

### LINKED ISSUE #3171: Nested config values in setting arrays and objects don't have environment variables expanded
Migrated from GitLab: https://gitlab.com/meltano/meltano/-/issues/3245

Originally created by @alxthm on 2022-02-15 09:06:57

---

Related to [#2366](https://gitlab.com/meltano/meltano/-/issues/2366) 

### What is the current *bug* behavior?

I was having trouble using environment variables in the `meltano.yml` file, and it seems they are only supported for top-level configuration options.

If I define `SOME_ENV_VAR=SOME_VALUE` in the `.env` file and I write this configuration yaml:
```yaml
plugins:
  extractors:
    - name: tap-spreadsheets-anywhere
      variant: ets
      pip_url: git+https://github.com/ets/tap-spreadsheets-anywhere.git
      config:
        test: $SOME_ENV_VAR
        tables:
          - path: $SOME_ENV_VAR
            name: my_table
```

Then this is the actual configuration used by meltano:
```
$ meltano config tap-spreadsheets-anywhere
{
  "tables": [
    {
      "path": "$SOME_ENV_VAR",
      "name": "my_table"
    }
  ],
  "test": "SOME_VALUE"
}
```

And the `tables.path` variable is not set.

### What is the expected *correct* behavior?

The environment variable `SOME_ENV_VAR` should expend to `SOME_VALUE` in the `tables` array as well, and we should have:
```
$ meltano config tap-spreadsheets-anywhere
{
  "tables": [
    {
      "path": "SOME_VALUE",
      "name": "my_table"
    }
  ],
  "test": "SOME_VALUE"
}
```

### Steps to reproduce

- add the configuration above in `meltano.yml`
- add the line `SOME_ENV_VAR=SOME_VALUE` to your `.env` file
- see the result with `meltano config tap-spreadsheets-anywhere`

### Relevant logs and/or screenshots
_Please use code blocks (\`\`\`) to format console output_

### Possible fixes
_If you can, link to the line of code that might be responsible for the problem or suggest a fix_

### Further regression test
_Ensure we automatically catch similar issues in the future_

- [ ] Write additional adequate test cases and submit test results
- [ ] Test results should be reviewed by a person from the team

### PR DESCRIPTION
**Pull request by edgarrmondragon**
Closes https://github.com/meltano/meltano/issues/3171

### CODE REVIEW

File: integration/example-library/meltano-expand-envvars-in-array/index.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+```shell
+meltano install
+```
+
+## Run the pipeline
+
+```shell
+ANIMAL_GROUP=birds meltano invoke tap-smoke-test
+ANIMAL_GROUP=reptiles meltano invoke tap-smoke-test
+ANIMAL_GROUP=mammals meltano invoke tap-smoke-test
```
Comment (edgarrmondragon) [inline integration/example-library/meltano-expand-envvars-in-array/index.md:18]: These expand this config

```yaml
- stream_name: $ANIMAL_GROUP
  input_filename: $MELTANO_PROJECT_ROOT/data/$ANIMAL_GROUP.jsonl
```

into different values and thus different pipelines :)

**Comment by ian-lewis-d** [comment]
This exactly fits my current case where I have data coming from several environments - all with the same format. As a result I _should_ be able to use the same pipeline without duplicating it. 

Releasing this fix would assist me hugely, and many others I'm sure.

**Review by WillDaSilva — APPROVED** [review]
We may want to use `t.Sequence` instead of `list`, but even without that change I'm very happy to see this change. Thanks @edgarrmondragon! (and sorry for taking so long to review it!)

File: src/meltano/core/utils/__init__.py [hunk 2]
```diff
@@ -477,7 +477,12 @@ def __init__(self, env_var: str):
     re.VERBOSE,
 )
 
-Expandable = t.TypeVar("Expandable", str, t.Mapping[str, "Expandable"])
+Expandable = t.TypeVar(
+    "Expandable",
+    str,
+    t.Mapping[str, "Expandable"],
+    t.List["Expandable"],
```
Comment (edgarrmondragon) [inline src/meltano/core/utils/__init__.py:484]:
```suggestion
    t.Sequence["Expandable"],
```

**Comment by edgarrmondragon** [comment]
> We may want to use `t.Sequence` instead of `list`, but even without that change I'm very happy to see this change. Thanks @edgarrmondragon! (and sorry for taking so long to review it!)

@WillDaSilva that makes sense 👍

Thanks for reviewing!

**Pull request merged**